ee/tabby-webserver/src/service/email.rs

```diff
@@ -76,23 +111,44 @@ pub async fn new_email_service(db: DbConn) -> Result<impl EmailService> {
 impl EmailService for EmailServiceImpl {
     async fn get_email_setting(&self) -> Result<Option<EmailSetting>> {
         let creds = self.db.read_email_setting().await?;
-        Ok(creds.map(Into::into))
+        let Some(creds) = creds else {
+            return Ok(None);
+        };
+        let creds = creds.try_into();
+        let Ok(creds) = creds else {
+            self.db.delete_email_setting().await?;
+            warn!("Email settings are corrupt, and have been deleted. Please reset them.");
+            return Ok(None);
+        };
+        Ok(Some(creds))
     }
 
     async fn update_email_setting(
         &self,
         smtp_username: String,
         smtp_password: Option<String>,
         smtp_server: String,
+        from_address: String,
+        encryption: Encryption,
+        auth_method: AuthMethod,
     ) -> Result<()> {
-        self.update_email_setting(
-            smtp_username.clone(),
-            smtp_password.clone(),
-            smtp_server.clone(),
-        )
-        .await?;
-        // TODO: make from address being configurable in EmailSettings.
-        *self.from.write().await = smtp_username.clone();
+        let input = EmailSettingInput {
+            smtp_username,
+            from_address,
+            smtp_server,
+        };
+        input.validate()?;
```

Comment:
Please follow the example in https://github.com/TabbyML/tabby/pull/1419 on *Input

validate should be called at graphql api layer to properly construct ValidationErrors

Reply:
This PR calls it in the service layer: https://github.com/TabbyML/tabby/pull/1419/files#diff-e0b0fc890e043329c44fcced75dee758af00d7637d9bd0a81e93cce1346935d4R17